File: benchmark.cpp
#include <vector>
#include <sys/time.h>
#include <boost/multi_array.hpp>

#include "adolc/adolc.h"
#include "numerical_helper_functions.hpp"

using namespace std;
using namespace boost;
typedef boost::multi_array<double,2> double_matrix;

/* return time in seconds */
struct timeval tv;
double mtime(void)
{
gettimeofday(&tv,NULL);
return (double)(tv.tv_sec + (tv.tv_usec / 1000000.));
}

ostream& operator<<(ostream& os, const double_matrix& A){
	for(int n = 0; n != (int) A.shape()[0]; ++n){
		for(int m = 0; m != (int) A.shape()[1]; ++m){
			os<<A[n][m]<<"\t";
		}
		os<<endl;
	}
	return os;
}

ostream& operator<<(ostream& os, const vector<double> x){
	for(int n = 0; n!= (int)x.size(); ++n){
		os<<x[n]<<"\t";
	}
	os<<endl;
	return os;
}

/* function f: \mathbb R^n \longrightarrow \mathbb R */
void double_speelpenning(const vector<double> &x, vector<double> &y){
	double tmp(1);
	for(int i = 0; i != x.size(); ++i){
		tmp*=x[i];
	}
	y[0] = tmp;
}
void adouble_speelpenning(const vector<adouble> &x, vector<adouble> &y){
	adouble tmp(1);
	for(int i = 0; i != x.size(); ++i){
		tmp*=x[i];
	}
	y[0] = tmp;
}

void double_matrix_vector_multiplication(const vector<double> &x, vector<double> &y){
	int N = x.size();
	int M = y.size();
	double_matrix A(extents[M][N]);
	for(int m = 0; m != M; ++m){
		for(int n = 0; n != N; ++n){
			A[m][n] = 1./N + (n==m);
		}
	}

	for(int m = 0; m != M; ++m){
		y[m] = 0.;
	}
	for(int m = 0; m != M; ++m){
		for(int n = 0; n != N; ++n){
			y[m]+= A[m][n]*x[n];
		}
	}
}
void adouble_matrix_vector_multiplication(const vector<adouble> &x, vector<adouble> &y){
	int N = x.size();
	int M = y.size();
	double_matrix A(extents[M][N]);
	for(int m = 0; m != M; ++m){
		for(int n = 0; n != N; ++n){
			A[m][n] = 1./N + (n==m);
		}
	}

	for(int m = 0; m != M; ++m){
		y[m] = 0.;
	}
	for(int m = 0; m != M; ++m){
		for(int n = 0; n != N; ++n){
			y[m]+= A[m][n]*x[n];
		}
	}
}


void benchmark(void (*adouble_f)(const vector<adouble>&, vector<adouble> &), void (*double_f)(const vector<double>&, vector<double> &),int N, int M, string message){
	double start_time;
	double runtime_taping;
	double runtime_adolc_function;
	double runtime_function;
	double runtime_gradient;
	double runtime_jacobian;

	vector<double> x(N);
	vector<double> g(N);
	vector<adouble> ax(N); 		/* active variables */
	vector<adouble> ay(M);
	vector<double> y(M);			/* y = f(x) */
	vector<double> y_normal(M);
	vector<double> directional_derivative(M);
	double **J = myalloc2(M,N);


	printf("%s", message.c_str());


	for(int n = 0; n!=N; ++n){
		 x[n] = 1./(n+1);
	}
	start_time = mtime();
	trace_on(1);
		for(int n = 0; n!=N; ++n) ax[n].operator<<=(x[n]);
		adouble_f(ax, ay);
		for(int m = 0; m!=M; ++m) ay[m]>>= y[m];
	trace_off();
	runtime_taping = mtime() - start_time;
	printf("Adolc\tfunction taping:\t........\telapsed time: %f\n",runtime_taping);


 	start_time = mtime();
	function(1, M, N, &x[0], &y[0]);
	runtime_adolc_function = mtime() - start_time;
	printf("Adolc\tfunction evaluation:\t%f\telapsed time: %f\n", y[0], runtime_adolc_function);


	start_time = mtime();
	double_f(x,y_normal);
	runtime_function = mtime() - start_time;
	printf("normal\tfunction evaluation:\t%f\telapsed time: %f\n",y_normal[0],runtime_function);


	start_time = mtime();
	jacobian(1, M, N, &x[0], J);
	runtime_jacobian = mtime() - start_time;
	printf("jacobian evaluation:\t\t........\telapsed time: %f\n", runtime_jacobian);

	if(M==1){
		start_time = mtime();
		gradient(1,N,x.data(),g.data());
		runtime_gradient = mtime() - start_time;
		printf("gradient evaluation:\t\t........\telapsed time: %f\n", runtime_gradient);
	}
	
}




int main(int argc, char* argv[]){
	int N;
	int M;

	if(argc == 3){
		N = atoi(argv[1]);
		M = atoi(argv[2]);
	}
	else{
		N = 10;
		M = 1;
	}
	cout<<"N="<<N<<" M="<<M<<endl;
	benchmark(adouble_speelpenning, double_speelpenning,N,1, "\n\nspeelpenning:\n");
	benchmark(adouble_matrix_vector_multiplication, double_matrix_vector_multiplication,N,M, "\n\nmatrix vector multiplication:\n");
	
	/* tape_doc(tag, m, n, x[n], y[m]) */   
// 	tape_doc(1, M, N, x, y);
	
	return EXIT_SUCCESS;
}

File: benchmark.py
#!/usr/bin/env python
import numpy as npy
from adolc import *
import time

def run_test(f,N,M, message='', print_derivatives=False):
	print message
	x = npy.array([1./(i+1) for i in range(N)])
	y = npy.zeros(M)
	ax = npy.array([adouble(0.) for i in range(N)])

	start_time = time.time()
	trace_on(1)
	for n in range(N):
		ax[n].is_independent(x[n])
	ay = f(ax)
	for m in range(M):
		y[m] = depends_on(ay[m])
	trace_off()
	tape_to_latex(1,x,y)
	runtime_taping = time.time() - start_time
	print 'Adolc\tfunction taping:\t........\telapsed time: %f'%runtime_taping

	start_time = time.time()
	y_adolc = function(1,x)
	runtime_adolc = time.time() - start_time
	print 'Adolc\tfunction evaluation:\t%f\telapsed time: %f'%(y_adolc[0],runtime_adolc)
	
	start_time = time.time()
	y_normal = f(x)
	runtime_normal = time.time() - start_time
	print 'normal\tfunction evaluation:\t%f\telapsed time: %f'%(y_normal[0],runtime_normal)

	if M==1:
		start_time = time.time()
		g = gradient(1,x)
		runtime_gradient = time.time() - start_time
		print 'gradient evaluation:\t\t........\telapsed time: %f'%runtime_gradient

	start_time = time.time()
	J = jacobian(1,x)
	runtime_jacobian = time.time() - start_time
	print 'jacobian evaluation:\t\t........\telapsed time: %f'%runtime_jacobian

	#start_time = time.time()
	#y_and_deriv = fos_forward(1,1,1,x,direction)
	#runtime_fos_forward = time.time() - start_time
	
	#if print_derivatives == True:
		 #print g
	#else:
		 #print ''
	#print 'direct. diff. evaluation:\t%f\telapsed time: %f'%(y_and_deriv['y'],runtime_fos_forward),
	#if print_derivatives == True:
		 #print  y_and_deriv['directional_derivative']
	#else:
		 #print ''
	#print ''
	#return y_normal


N = 500
M = 500
direction = npy.zeros(N)
direction[0] = 1.


#def f(avec):
	#a = avec[0]*avec[-1]
	#c =avec[0]*avec[0]
	#if isinstance(a,badouble):
		#print avec[0].location
		#print avec[1].location
		#print a.location
		#print c.location
	#return a
#y = run_test(f,x,direction)


#py_tape_doc(1,x,npy.array([y]))

#def f(avec):
	#return npy.sum(avec)
#run_test(f,x,direction)

#print "N=%d   M=%d"%(N,M)

#def f(avec):
	#return npy.array([npy.prod(avec)])
#run_test(f,N,1,'\n\nspeelpenning')


A = npy.zeros((M,N))
A[:] = [[ 1./N +(n==m) for n in range(N)] for m in range(M)]

def f(x):
	global A
	return npy.dot(A,x)
run_test(f,N,M,'\n\nmatrix vector multiplication')

